### PR TITLE
limit the range of numeric predictions

### PR DESCRIPTION
**Pull request by topepo**
A simple post-processor that can cap the range of predicted values form regression models.

### CODE REVIEW

File: R/bound_prediction.R [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#' Truncate a numeric prediction column
+#'
+#' For user-defined lower_limit and/or upper_limit bound, ensure that the values in the
```
Comment (simonpcouch) [inline R/bound_prediction.R:3]:
```suggestion
#' For user-defined `lower_limit` and/or `upper_limit` bound, ensure that the values in the
```

File: R/bound_prediction.R [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#' `.pred` column are coerced to these bounds.
+#'
+#' @param x A data frame that contains a numeric column named `.pred`.
+#' @param lower_limit,upper_limit Single numerics (or `NA`) that define
+#' constrains on `.pred`.
```
Comment (simonpcouch) [inline R/bound_prediction.R:8]:
```suggestion
#' constraints on `.pred`.
```

File: R/bound_prediction.R [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#' @return `x` with potentially adjusted values.
+#' @examples
+#' data(solubility_test, package = "yardstick")
+#'
+#' names(solubility_test) <- c("solubility", ".pred")
+#'
+#' bound_prediction(solubility_test, lower_limit = -1)
+#' @export
+bound_prediction <- function(x, lower_limit = -Inf, upper_limit = Inf,
+                             call = rlang::caller_env()) {
```
Comment (simonpcouch) [inline R/bound_prediction.R:19]:
```suggestion
                             call = rlang::current_env()) {
```

The calling environment from where this is evaluated is likely one step up from where we want it. If you'd like to call `bound_predictions()` yourself and see `Error in bound_predictions()`, we'll need to bump this down one frame. 

This change will affect snaps. :)

File: tests/testthat/test-bound-prediction.R [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    error = TRUE)
+
+  sol <- solubility_test %>% set_names(c("solubility", ".pred"))
+
+  expect_equal(bound_prediction(sol), sol)
+  expect_equal(bound_prediction(sol, lower_limit = NA), sol)
+
+  res_1 <- bound_prediction(sol, lower_limit = -1)
+  expect_true(all(res_1$.pred[res_1$.pred < -1] == -1))
+  expect_true(all(res_1$.pred[res_1$.pred >= -1] == res_1$.pred[res_1$.pred >= -1]))
```
Comment (simonpcouch) [inline tests/testthat/test-bound-prediction.R:24]:
```suggestion
  expect_true(all(res_1$.pred[res_1$.pred >= -1] == sol$.pred[sol$.pred >= -1]))
```

I think? We just want to test that values are unaffected?

File: tests/testthat/test-bound-prediction.R [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    error = TRUE)
+
+  sol <- solubility_test %>% set_names(c("solubility", ".pred"))
+
+  expect_equal(bound_prediction(sol), sol)
+  expect_equal(bound_prediction(sol, upper_limit = NA), sol)
+
+  res_1 <- bound_prediction(sol, upper_limit = -1)
+  expect_true(all(res_1$.pred[res_1$.pred > -1] == -1))
+  expect_true(all(res_1$.pred[res_1$.pred <= -1] == res_1$.pred[res_1$.pred <= -1]))
```
Comment (simonpcouch) [inline tests/testthat/test-bound-prediction.R:52]:
```suggestion
  expect_true(all(res_1$.pred[res_1$.pred <= -1] == sol$.pred[sol$.pred <= -1]))
```

File: tests/testthat/test-bound-prediction.R (hunk 4 above)
Comment (simonpcouch) [reply tests/testthat/test-bound-prediction.R:24]: In that, `expect_equal()` might give more informative test failures than `==`.

File: tests/testthat/test-bound-prediction.R [hunk 6]
```diff
@@ -0,0 +1,55 @@
+test_that("lower_limit bounds for numeric predictions", {
+  skip_if_not_installed("modeldata")
+  library(dplyr)
+  library(rlang)
+  data("solubility_test", package = "modeldata")
+  tune2 <- function() call("tune", "test")
```
Comment (simonpcouch) [inline tests/testthat/test-bound-prediction.R:6]: Is this function ever going to have to deal with `tune()` values. I guess I'm using my mental model from parsnip here, where upstream packages take care of `tune()` calls before we send specs off to `fit()`.

File: R/bound_prediction.R [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  if (!any(names(x) == ".pred")) {
+    cli::cli_abort("The argument {.arg x} should have a column named {.code .pred}",
```
Comment (simonpcouch) [inline R/bound_prediction.R:21]:
```suggestion
    cli::cli_abort("The argument {.arg x} should have a column named {.code .pred}.",
```

File: tests/testthat/Rplots.pdf
Comment (simonpcouch) [inline tests/testthat/Rplots.pdf:1]: Not sure what this is?

File: NEWS.md [hunk 8]
```diff
@@ -1,5 +1,7 @@
 # probably (development version)
 
+* A new function `bound_prediction()` is available to constrain the values of a numeric prediction.
```
Comment (simonpcouch) [inline NEWS.md:3]:
```suggestion
* A new function `bound_prediction()` is available to constrain the values of a numeric prediction (#142).
```

**Review by simonpcouch — COMMENTED** [review]
Looking good😎

In general, I think the arguments to this function would be well-served by the rlang type checkers. There are a good few edge cases (>1-length numeric to upper_limit, non-data-frame x) that would be caught by those utilities, and their messages are quite informative.

File: tests/testthat/test-bound-prediction.R (hunk 6 above)
Comment (topepo) [reply tests/testthat/test-bound-prediction.R:6]: It might erroneously receive that from users. I'd err on the side of over-checking

File: tests/testthat/Rplots.pdf
Comment (topepo) [reply tests/testthat/Rplots.pdf:1]: Detritus. I'll remove it.

**Comment by topepo** [comment]
>  I think the arguments to this function would be well-served by the rlang type checkers. 

Just added them. This would be a good set of issues for TDD 🥳

File: tests/testthat/test-bound-prediction.R (hunk 4 above)
Comment (topepo) [reply tests/testthat/test-bound-prediction.R:24]: omg 😬

File: tests/testthat/test-bound-prediction.R (hunk 6 above)
Comment (topepo) [reply tests/testthat/test-bound-prediction.R:6]: I added an error for these values (instead of just passing)
